frontend/apps/reader/modules/readertoc.lua

```diff
@@ -168,25 +168,35 @@ function ReaderToc:validateAndFixToc()
             nb_bogus = nb_bogus + 1
             -- See how many pages we'd need fixing on either side
             local nb_prev = 0
+            local nb_prev_main = 0
             for j = i-1, first, -1 do
                 local ppage = toc[j].fixed_page or toc[j].page
                 if ppage <= page then
                     break
                 else
                     nb_prev = nb_prev + 1
+                    if self.ui.document:getPageFlow(ppage) == 0 then
+                        nb_prev_main = nb_prev_main + 1
+                    end
                 end
             end
-            local nb_next = 1
-            for j = i+1, last do
+            local nb_next = 0
+            local nb_next_main = 0
+            for j = i, last do
                 local npage = toc[j].fixed_page or toc[j].page
                 if npage >= cur_page then
                     break
                 else
                     nb_next = nb_next + 1
+                    if self.ui.document:getPageFlow(npage) == 0 then
+                        nb_next_main = nb_next_main + 1
+                    end
                 end
             end
-            logger.dbg("BOGUS TOC:", i, page, ">", i-1, cur_page, "-", nb_prev, nb_next)
-            if nb_prev <= nb_next then -- less changes when fixing previous pages
+            logger.dbg("BOGUS TOC:", i, page, "<", i-1, cur_page, "-", nb_prev, nb_next)
+            -- Note: by comparing only the entries that belong to the main (linear) flow
+            -- we give priority to moving non-linear bogus entries
+            if nb_prev_main <= nb_next_main then -- less changes when fixing previous pages
```

Comment:
No issue when only TOC items in non-linear flows are misordered ? Our TOC code expects ALL TOC entries to be ordered, so if this is screwed by multiple entries in non-linear parts, skipped there, it will be an issue.

Reply:
Nothing has changed here, except that:
* The displayed page in the TOC is the local page.
* The decision on whether to fix the bogus pages "before" or "after" is done based on the number of linear entries that need to be fixed, but all linear and non-linear are fixed. The idea is that we prefer fixing non-linear entries than linear ones.

At least that is the intent.